src/incubator/TransitionAnimator/index.tsx

```diff
@@ -0,0 +1,140 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+import React, {PropsWithChildren, useEffect, useCallback, useImperativeHandle} from 'react';
+import {View as RNView, LayoutChangeEvent} from 'react-native';
+import Animated, {useSharedValue, useAnimatedStyle, withSpring, withTiming, runOnJS} from 'react-native-reanimated';
+import View, {ViewProps} from '../../components/view';
+import {forwardRef, ForwardRefInjectedProps} from '../../commons/new';
+import useHiddenLocations, {HiddenLocation} from './useHiddenLocations';
+const AnimatedView = Animated.createAnimatedComponent(View);
+export {HiddenLocation as TransitionLocation};
+
+const DEFAULT_ANIMATION_VELOCITY = 300;
+const DEFAULT_ANIMATION_CONFIG = {velocity: DEFAULT_ANIMATION_VELOCITY, damping: 18, stiffness: 300, mass: 0.4};
+
+export type TransitionAnimationEndType = 'in' | 'out';
+
+export interface TransitionAnimatorProps extends ViewProps {
+  /**
+   * Callback to the animation end.
+   */
+  onAnimationEnd?: (animationType: TransitionAnimationEndType) => void;
+  /**
+   * If this is given there will be an enter animation from this location.
+   */
+  enterAnimationLocation?: HiddenLocation;
+  /**
+   * If this is given there will be an exit animation in this location.
+   */
+  exitAnimationLocation?: HiddenLocation;
+}
+
+type Props = PropsWithChildren<TransitionAnimatorProps> & ForwardRefInjectedProps;
+interface Statics {
+  animateOut: () => void;
+}
+
+const TransitionAnimator = (props: Props) => {
+  const {
+    onAnimationEnd,
+    enterAnimationLocation,
+    exitAnimationLocation,
+    forwardedRef,
+    style: propsStyle,
+    onLayout: propsOnLayout,
+    ...others
+  } = props;
+  const containerRef = React.createRef<RNView>();
+  const {onLayout: hiddenLocationOnLayout, hiddenLocations} = useHiddenLocations({containerRef});
+  const visible = useSharedValue<boolean>(!enterAnimationLocation);
+
+  const getLocation = (location?: HiddenLocation) => {
+    if (location === 'left' || location === 'right') {
```

Comment:
I personally like to use includes in these cases (`['left', 'right'].includes(location)`)
I found it shorter and more readable. 

In general this whole method can be shorten to 

```
return {
  x: ['left', 'right'].includes(location) ? hiddenLocations[location] : 0,
  y: ['top', 'bottom'].includes(location) ? hiddenLocations[location] : 0,
}
```